<a href="https://colab.research.google.com/github/Pratham5321/Movie-Sentiment-Analysis/blob/main/word2vec_approach.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import recquired libraries

In [1]:
import pandas as pd
import numpy as np

In [2]:
# download the data from the github

!wget https://raw.githubusercontent.com/laxmimerit/All-CSV-ML-Data-Files-Download/master/IMDB-Dataset.csv

--2026-05-12 14:57:45--  https://raw.githubusercontent.com/laxmimerit/All-CSV-ML-Data-Files-Download/master/IMDB-Dataset.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 66212309 (63M) [text/plain]
Saving to: ‘IMDB-Dataset.csv’

IMDB-Dataset.csv    100%[===================>]  63.14M   298MB/s    in 0.2s    

2026-05-12 14:57:47 (298 MB/s) - ‘IMDB-Dataset.csv’ saved [66212309/66212309]



In [3]:
# import the dataset

df=pd.read_csv("IMDB-Dataset.csv")
df.head()

,review,sentiment
0,One of the other reviewers has mentioned that ...,positive
1,A wonderful little production. <br /><br />The...,positive
2,I thought this was a wonderful way to spend ti...,positive
3,Basically there's a family where a little boy ...,negative
4,"Petter Mattei's ""Love in the Time of Money"" is...",positive


In [4]:
# check for the summary of the data

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     50000 non-null  object
 1   sentiment  50000 non-null  object
dtypes: object(2)
memory usage: 781.4+ KB


- we have two columns first is the review which contain the review about the movie.

- and in sentiment column we have the two values positive and negative values .

In [5]:
# check for the null values

df.isna().sum()

,0
review,0
sentiment,0


- there is no any null values in the whole dataset.

In [6]:
# check for the duplicates

df.duplicated().sum()

np.int64(418)

- there is an 418 duplicated rows in the dataset.

In [7]:
# here we clean the duplicated

df.drop_duplicates(inplace=True)

# check for the duplicated rows

df.duplicated().sum()

np.int64(0)

In [8]:
# here we encode the target values

df["sentiment"]=df["sentiment"].map({"positive":1 , "negative":0 })
df

,review,sentiment
0,One of the other reviewers has mentioned that ...,1
1,A wonderful little production. <br /><br />The...,1
2,I thought this was a wonderful way to spend ti...,1
3,Basically there's a family where a little boy ...,0
4,"Petter Mattei's ""Love in the Time of Money"" is...",1
...,...,...
49995,I thought this movie did a down right good job...,1
49996,"Bad plot, bad dialogue, bad acting, idiotic di...",0
49997,I am a Catholic taught in parochial elementary...,0
49998,I'm going to have to disagree with the previou...,0


# now we preprocess the data

In [9]:
# import the recuired libraries

import nltk
import string
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

nltk.download("stopwords")
nltk.download("punkt_tab")

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

In [10]:
# tokenize the review column

df["review_token"]=df["review"].apply(word_tokenize)
df

,review,sentiment,review_token
0,One of the other reviewers has mentioned that ...,1,"[One, of, the, other, reviewers, has, mentione..."
1,A wonderful little production. <br /><br />The...,1,"[A, wonderful, little, production, ., <, br, /..."
2,I thought this was a wonderful way to spend ti...,1,"[I, thought, this, was, a, wonderful, way, to,..."
3,Basically there's a family where a little boy ...,0,"[Basically, there, 's, a, family, where, a, li..."
4,"Petter Mattei's ""Love in the Time of Money"" is...",1,"[Petter, Mattei, 's, ``, Love, in, the, Time, ..."
...,...,...,...
49995,I thought this movie did a down right good job...,1,"[I, thought, this, movie, did, a, down, right,..."
49996,"Bad plot, bad dialogue, bad acting, idiotic di...",0,"[Bad, plot, ,, bad, dialogue, ,, bad, acting, ..."
49997,I am a Catholic taught in parochial elementary...,0,"[I, am, a, Catholic, taught, in, parochial, el..."
49998,I'm going to have to disagree with the previou...,0,"[I, 'm, going, to, have, to, disagree, with, t..."


In [11]:
# check the tokens size in the first review

len(df["review_token"][0])

380

In [12]:
# convert all tokens in the small form

df["review_token"]=df["review_token"].apply(lambda x:[i.lower() for i in x])
df.head()

,review,sentiment,review_token
0,One of the other reviewers has mentioned that ...,1,"[one, of, the, other, reviewers, has, mentione..."
1,A wonderful little production. <br /><br />The...,1,"[a, wonderful, little, production, ., <, br, /..."
2,I thought this was a wonderful way to spend ti...,1,"[i, thought, this, was, a, wonderful, way, to,..."
3,Basically there's a family where a little boy ...,0,"[basically, there, 's, a, family, where, a, li..."
4,"Petter Mattei's ""Love in the Time of Money"" is...",1,"[petter, mattei, 's, ``, love, in, the, time, ..."


In [13]:
# now remove the punctuations

df["review_token"]=df["review_token"].apply(lambda x : [i for i in x if i not in string.punctuation])
df.head()

,review,sentiment,review_token
0,One of the other reviewers has mentioned that ...,1,"[one, of, the, other, reviewers, has, mentione..."
1,A wonderful little production. <br /><br />The...,1,"[a, wonderful, little, production, br, br, the..."
2,I thought this was a wonderful way to spend ti...,1,"[i, thought, this, was, a, wonderful, way, to,..."
3,Basically there's a family where a little boy ...,0,"[basically, there, 's, a, family, where, a, li..."
4,"Petter Mattei's ""Love in the Time of Money"" is...",1,"[petter, mattei, 's, ``, love, in, the, time, ..."


In [14]:
# check the tokens size in the first review

len(df["review_token"][0])

324

In [15]:
# now remove the stopwords form the vocabulary

stopword=stopwords.words("english")


df["review_token"]=df["review_token"].apply(lambda x: [i for i in x if i not in stopword])
df.head()

,review,sentiment,review_token
0,One of the other reviewers has mentioned that ...,1,"[one, reviewers, mentioned, watching, 1, oz, e..."
1,A wonderful little production. <br /><br />The...,1,"[wonderful, little, production, br, br, filmin..."
2,I thought this was a wonderful way to spend ti...,1,"[thought, wonderful, way, spend, time, hot, su..."
3,Basically there's a family where a little boy ...,0,"[basically, 's, family, little, boy, jake, thi..."
4,"Petter Mattei's ""Love in the Time of Money"" is...",1,"[petter, mattei, 's, ``, love, time, money, ''..."


In [16]:
# check the tokens size in the first review

len(df["review_token"][0])

182

In [17]:
# now we preprocessed the data we are ready apply the word2vec

! pip install gensim

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.9/27.9 MB 82.8 MB/s eta 0:00:00


# genism

- Word2Vec
- Doc2Vec
- FastText
- topic modeling
in NLP.

In [18]:
import gensim
from gensim.models import Word2Vec

In [19]:
# apply the word 2 vec model

model=Word2Vec(
    sentences=df["review_token"],
    vector_size=100,
    window=5,
    min_count=2
)

In [20]:
model.wv["good"]

array([ 0.69798493, -1.4904644 , -0.12182393, -0.9828974 , -2.1688652 ,
       -0.08248829,  2.3575404 ,  1.6778963 , -2.2871451 , -1.1064494 ,
        1.8838257 ,  1.4682839 ,  0.84837353,  1.72316   , -1.5098847 ,
        1.6497029 ,  1.7880158 ,  0.3855752 ,  0.88287187, -1.9715307 ,
        0.20472786, -0.6843082 ,  1.5451547 , -2.0262709 ,  2.3443642 ,
       -0.01095021, -1.8661823 ,  1.1848464 , -0.40488625, -4.0486846 ,
        0.8006156 ,  0.8828708 ,  0.7137256 ,  2.608707  ,  0.6201632 ,
       -1.4120896 ,  1.1651621 ,  0.37741253,  1.0480411 ,  1.0208019 ,
       -0.10406371, -2.4011602 , -0.51248664, -2.9214642 , -0.71668684,
       -1.470734  ,  0.73045003, -0.27181014,  0.39145806,  0.46326756,
        2.6711843 ,  0.99475044, -0.49363008,  2.8294303 ,  2.0852003 ,
       -0.21328336, -0.37095216, -1.3483118 ,  1.3073688 ,  1.8156848 ,
       -0.2997254 , -1.035748  ,  1.4622085 ,  1.6137326 , -0.3476174 ,
        1.4673569 , -0.69423354, -0.7995391 ,  0.98848027, -0.64

In [21]:
model.wv.most_similar("production")

[('product', 0.6013288497924805),
 ('transfer', 0.5512717962265015),
 ('budgets', 0.532051146030426),
 ('project', 0.5205557346343994),
 ('quality', 0.5138739347457886),
 ('no-quality', 0.5137723684310913),
 ('editing', 0.5137428641319275),
 ('directing', 0.5070787072181702),
 ('cinematography', 0.5062739849090576),
 ('budget', 0.5058554410934448)]

In [37]:
# convert the each vocabulary to the vector

def vectorizer(vocabulary):

  vector=[]

  for word in vocabulary:
    if word in model.wv:
      vector.append(model.wv[word])

    if len(vector)==0:
      return np.zeros(model.vector_size)

  return np.mean(vector,axis=0)

In [38]:
# apply the function on the tokens

df["token_vector"]=df["review_token"].apply(vectorizer)
df.head()

,review,sentiment,review_token,token_vector
0,One of the other reviewers has mentioned that ...,1,"[one, reviewers, mentioned, watching, 1, oz, e...","[-0.011204327, 0.3252881, 0.13827884, -0.24604..."
1,A wonderful little production. <br /><br />The...,1,"[wonderful, little, production, br, br, filmin...","[0.040121395, 0.48685586, 0.39134598, 0.156993..."
2,I thought this was a wonderful way to spend ti...,1,"[thought, wonderful, way, spend, time, hot, su...","[0.11122498, 0.23777314, 0.06729436, -0.147772..."
3,Basically there's a family where a little boy ...,0,"[basically, 's, family, little, boy, jake, thi...","[-0.06635332, 0.2062372, 0.13988072, 0.141878,..."
4,"Petter Mattei's ""Love in the Time of Money"" is...",1,"[petter, mattei, 's, ``, love, time, money, ''...","[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ..."


In [52]:
# now we create the independant and dependant variables

x=df["token_vector"].tolist()
y=df["sentiment"].tolist()

In [53]:
# now split into the train and test sets

from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

- here we preprossed the data from convert the vocabulary into the vectors

# Now we train the machine learning model

In [54]:
# first we train on the logistic regression

from sklearn.linear_model import LogisticRegression

linear_model=LogisticRegression()
linear_model.fit(x_train,y_train)


LogisticRegression()

In [55]:
# now evaluate the models performance on test set

from sklearn.metrics import accuracy_score,classification_report

y_pred=linear_model.predict(x_test)

print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

0.8557023293334678
              precision    recall  f1-score   support

           0       0.86      0.85      0.85      4939
           1       0.85      0.86      0.86      4978

    accuracy                           0.86      9917
   macro avg       0.86      0.86      0.86      9917
weighted avg       0.86      0.86      0.86      9917



In [56]:
# now evaluate the models performance on train set

from sklearn.metrics import accuracy_score,classification_report

y_pred=linear_model.predict(x_train)

print(accuracy_score(y_train,y_pred))
print(classification_report(y_train,y_pred))

0.8604563216941888
              precision    recall  f1-score   support

           0       0.86      0.86      0.86     19759
           1       0.86      0.86      0.86     19906

    accuracy                           0.86     39665
   macro avg       0.86      0.86      0.86     39665
weighted avg       0.86      0.86      0.86     39665



In [71]:
# here we try the hyperparameter tuning

from sklearn.model_selection import GridSearchCV

logi_model=LogisticRegression()

param_grid={"C":[0.01,0.1,1.0],
            "solver":["liblinear","saga"],
            "penalty":["l1","l2"]}


grid=GridSearchCV(logi_model,param_grid,cv=5,scoring="accuracy")
grid.fit(x_train,y_train)


GridSearchCV(cv=5, estimator=LogisticRegression(),
             param_grid={'C': [0.01, 0.1, 1.0], 'penalty': ['l1', 'l2'],
                         'solver': ['liblinear', 'saga']},
             scoring='accuracy')

In [78]:
# train and evalueate with best estamators

logi_model=LogisticRegression(**grid.best_params_)

logi_model.fit(x_train,y_train)

y_pred=logi_model.predict(x_test)

print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

0.855803166280125
              precision    recall  f1-score   support

           0       0.86      0.85      0.85      4939
           1       0.85      0.86      0.86      4978

    accuracy                           0.86      9917
   macro avg       0.86      0.86      0.86      9917
weighted avg       0.86      0.86      0.86      9917



- there is no such a big diffrence between the hyperparameter and the custom parameters

#  the train and test set results are similar so the model is well fitted and gives 86% accuracy on both sets .